src/api2/wallets-manager.js

```diff
@@ -0,0 +1,502 @@
+
+    const endpoint = `/wallets/${verificationWalletId}/verification`
+    return this._apiCaller.post(endpoint)
+  }
+
+  /**
+   * Recover a wallet using the recovery seed.
+   *
+   * @param {string} email User's email.
+   * @param {string} recoverySeed User's recovery seed.
+   * @param {string} newPassword Desired password.
+   *
+   * @return {Wallet} New wallet.
+   */
+  async recovery (email, recoverySeed, newPassword) {
+    const { data: kdfParams } = await this.getKdfParams(email, true)
+
+    const recoveryWallet = Wallet.fromRecoverySeed(
+      kdfParams,
+      kdfParams.salt,
+      email,
+      recoverySeed
+    )
+
+    const newMainWallet = Wallet.generate(email)
+    const encryptedNewMainWallet = newMainWallet.encrypt(kdfParams, newPassword)
+
+    const newSecondFactorWallet = Wallet.generate(email)
+    const encryptedSecondFactorWallet = newSecondFactorWallet.encrypt(
+      kdfParams,
+      newPassword
+    )
+
+    const accountId = await this._getAccountIdByRecoveryId(recoveryWallet.id)
+    const signers = await this._getSigners(accountId)
+    const signerRoleId = await this._getDefaultSignerRole()
+
+    const tx = makeChangeSignerTransaction({
+      newPublicKey: newMainWallet.accountId,
+      signers,
+      signingKeypair: recoveryWallet.keypair,
+      soucreAccount: accountId,
+      signerRoleId
+    })
+
+    const oldWallet = this._apiCaller.wallet
+    this._apiCaller.useWallet(recoveryWallet)
+
+    const endpoint = `/wallets/${recoveryWallet.id}`
+
+    try {
+      await this._apiCaller.putWithSignature(endpoint, {
+        data: {
+          type: 'wallet',
+          id: encryptedNewMainWallet.id,
+          attributes: {
+            email,
+            account_id: encryptedNewMainWallet.accountId,
+            salt: encryptedNewMainWallet.salt,
+            keychain_data: encryptedNewMainWallet.keychainData
+          },
+          relationships: {
+            transaction: {
+              data: {
+                type: 'transaction',
+                id: '1'
+              }
+            },
+            kdf: {
+              data: {
+                type: kdfParams.type,
+                id: kdfParams.id
+              }
+            },
+            factor: {
+              data: {
+                type: 'password',
+                id: encryptedNewMainWallet.id
+              }
+            }
+          }
+        },
+        included: [
+          {
+            type: 'transaction',
+            id: '1',
+            attributes: { envelope: tx }
+          },
+          {
+            id: encryptedNewMainWallet.id,
+            type: 'password',
+            attributes: {
+              account_id: encryptedSecondFactorWallet.accountId,
+              keychain_data: encryptedSecondFactorWallet.keychainData,
+              salt: encryptedSecondFactorWallet.salt
+            }
+          }
+        ]
+      })
+    } finally {
+      this._apiCaller.useWallet(oldWallet)
```

Comment:
mb better to make `putWithSignature` also accept the wallet as a 3rd param?

Reply:
with the default value

Reply:
@max-nes

Reply:
Discussed IRL